File: calculadora.py
import math

def ler_valores():
    try:
        tipo = input("Tipo de termo (p para polinomial, e para exponencial): ").strip().lower()
        coef = float(input("Digite o coeficiente: "))
    except ValueError:
        print("Entrada inválida! Assumindo coeficiente 0.")
        coef = 0
    
    if tipo == 'p':
        try:
            exp = float(input("Digite o expoente: "))
            return {"tipo": "polinomial", "coeficiente": coef, "expoente": exp}
        except ValueError:
            print("Expoente inválido! Usando 1.")
            return {"tipo": "polinomial", "coeficiente": coef, "expoente": 1}
    elif tipo == 'e':
        base_input = input("Digite a base da exponencial (pode ser 'e' para base natural): ").strip().lower()
        try:
            base = math.e if base_input == 'e' else float(base_input)
            if base <= 0:
                print("Base inválida! Usando 1.")
                base = 1
            return {"tipo": "exponencial", "coeficiente": coef, "base": base}
        except ValueError:
            print("Base inválida! Usando 1.")
            return {"tipo": "exponencial", "coeficiente": coef, "base": 1}
    else:
        print("Tipo inválido! Assumindo polinomial com expoente 1.")
        return {"tipo": "polinomial", "coeficiente": coef, "expoente": 1}

def imprime_funcao(funcao):
    for i, termo in enumerate(funcao):
        coef = termo["coeficiente"]
        if termo["tipo"] == "polinomial":
            exp = termo["expoente"]
            coef_str = f"{int(coef)}" if coef.is_integer() else f"{coef:.4f}"
            exp_str = f"{int(exp)}" if exp.is_integer() else f"{exp:.4f}"
            print(f"Termo {i + 1}: {coef_str} * x^{exp_str}")
        else:
            base = termo["base"]
            base_str = "e" if abs(base - math.e) < 1e-6 else f"{base:.4f}"
            coef_str = f"{int(coef)}" if coef.is_integer() else f"{coef:.4f}"
            print(f"Termo {i + 1}: {coef_str} * {base_str}^x")

def derivada(funcao):
    derivada_funcao = []
    for termo in funcao:
        if termo["tipo"] == "polinomial":
            if termo["expoente"] == 0:
                derivada_funcao.append({
                    "tipo": "polinomial",
                    "coeficiente": 0,
                    "expoente": 0
                })
            else:
                derivada_funcao.append({
                    "tipo": "polinomial",
                    "coeficiente": termo["coeficiente"] * termo["expoente"],
                    "expoente": termo["expoente"] - 1
                })
        elif termo["tipo"] == "exponencial":
            derivada_funcao.append({
                "tipo": "exponencial",
                "coeficiente": termo["coeficiente"] * math.log(termo["base"]),
                "base": termo["base"]
            })
    return derivada_funcao

def avaliar(funcao, x):
    resultado = 0
    for termo in funcao:
        if termo["tipo"] == "polinomial":
            resultado += termo["coeficiente"] * (x ** termo["expoente"])
        elif termo["tipo"] == "exponencial":
            resultado += termo["coeficiente"] * (termo["base"] ** x)
    return resultado

def encontrar_pontos_criticos(derivada_funcao, inicio=-10, fim=10, passo=0.001, tolerancia=1e-8):
    pontos_criticos = []
    x = inicio
    y_anterior = avaliar(derivada_funcao, x)
    while x <= fim:
        y = avaliar(derivada_funcao, x)
        if abs(y) < tolerancia or (y * y_anterior < 0): 
            if round(x, 3) not in pontos_criticos:
                pontos_criticos.append(round(x, 3))
        y_anterior = y
        x += passo
    return pontos_criticos

try:
    n = int(input("Digite a quantidade de termos: "))
    if n <= 0:
        print("Número de termos deve ser positivo!")
        n = 1
except ValueError:
    print("Entrada inválida! Usando 1 termo.")
    n = 1

funcao = []
for i in range(n):
    funcao.append(ler_valores())
print("\nFunção original:")
imprime_funcao(funcao)

primeira_derivada = derivada(funcao)
print("\nPrimeira derivada da função:")
imprime_funcao(primeira_derivada)

pontos = encontrar_pontos_criticos(primeira_derivada)
print("\nPontos críticos encontrados (f'(x) ≈ 0):")
print(pontos if pontos else "Nenhum ponto crítico encontrado.")

segunda_derivada = derivada(primeira_derivada)
print("\nSegunda derivada da função:")
imprime_funcao(segunda_derivada)

print("\nClassificação dos pontos críticos:")
for x in pontos:
    segunda = avaliar(segunda_derivada, x)
    if segunda > 0:
        print(f"x = {x}: Mínimo local")
    elif segunda < 0:
        print(f"x = {x}: Máximo local")
    else:
        print(f"x = {x}: Ponto de inflexão ou indeciso")

File: calculadora.js
const prompt = require("prompt-sync")();

function ler_valores() {
    let tipo = prompt("Tipo de termo (p para polinomial, e para exponencial): ");
    let coef = Number(prompt("Digite o coeficiente: "));
    
    if (isNaN(coef)) {
        console.log("Coeficiente inválido! Usando 0.");
        coef = 0;
    }

    if (tipo === "p") {
        let exp = Number(prompt("Digite o expoente: "));
        if (isNaN(exp)) {
            console.log("Expoente inválido! Usando 1.");
            exp = 1;
        }
        return { tipo: "polinomial", coef, exp };
    } else if (tipo === "e") {
        let base_input = prompt("Digite a base da exponencial (ou 'e' para base natural): ");
        let base = base_input === "e" ? Math.E : Number(base_input);
        if (isNaN(base) || base <= 0) {
            console.log("Base inválida! Usando 1.");
            base = 1;
        }
        return { tipo: "exponencial", coef, base };
    } else {
        console.log("Tipo inválido! Assumindo polinomial com expoente 1.");
        return { tipo: "polinomial", coef, exp: 1 };
    }
}

function imprime_funcao(funcao) {
    funcao.forEach((termo, i) => {
        if (termo.tipo === "polinomial") {
            console.log(`Termo ${i + 1}: ${termo.coef} * x^${termo.exp}`);  
        } else {
            console.log(`Termo ${i + 1}: ${termo.coef.toFixed(4)} * ${termo.base.toFixed(4)}^x`);  
        }
    });
}

function derivada(funcao) {
    return funcao.map((termo) => {
        if (termo.tipo === "polinomial") {
            return termo.exp === 0 
            ? { tipo: "polinomial", coef: 0, exp: 0 }
            : { tipo: "polinomial", coef: termo.coef * termo.exp, exp: termo.exp - 1 };
        } else if (termo.tipo === "exponencial") {
            return {
                tipo: "exponencial",
                coef: termo.coef * Math.log(termo.base),
                base: termo.base
            }
        }
    });
}

function avaliar(funcao, x) {
    let resultado = 0;
    funcao.forEach(termo => {
        if (termo.tipo === "polinomial") {
            resultado += termo.coef * (x ** termo.exp);
        } else {
            resultado += termo.coef * (termo.base ** x); 
        }
    });
    return resultado;
}

function encontrar_pontos_criticos(primeiraDerivada, inicio = -10, fim = 10, passo = 0.01, tolerancia = 1e-3) {
    let pontos_criticos = [];
    let x = inicio;
    let y_anterior = avaliar(primeiraDerivada, x);
    while (x <= fim) {
        let y = avaliar(primeiraDerivada, x);
        if (Math.abs(y) < tolerancia || y * y_anterior < 0) { // Detecta mudança de sinal
            let ponto = Number(x.toFixed(3));
            if (!pontos_criticos.includes(ponto)) { // Evita duplicatas
                pontos_criticos.push(ponto);
            }
        }
        y_anterior = y;
        x += passo;
    }
    return pontos_criticos;
}

let n = Number(prompt("Digite a quantidade de termos: "));
if (n <= 0) {
    console.log("Número de termos deve ser positivo!");
    n = 1;
} else if (isNaN(n)) {
    console.log("Entrada inválida! Usando 1 termo.");
    n = 1;
}

let funcao = [];
for (let i = 0; i < n; i++) {
    funcao.push(ler_valores());
}
console.log("\nFunção original:");
imprime_funcao(funcao);

let primeiraDerivada = derivada(funcao);
console.log("\nPrimeira derivada da função:");
imprime_funcao(primeiraDerivada);

let pontos = encontrar_pontos_criticos(primeiraDerivada);
console.log(`\nPontos críticos encontrados (f'(x) ≈ 0):`);
console.log(pontos.length > 0 ? pontos : "Nenhum ponto crítico encontrado.");

let segundaDerivada = derivada(primeiraDerivada);
console.log("\nSegunda derivada da função:");
imprime_funcao(segundaDerivada);

console.log("\nClassificação dos pontos críticos:");
pontos.forEach((x) => {
    let segunda = avaliar(segundaDerivada, x);
    if (segunda > 0) {
        console.log(`x = ${x}: Mínimo local`);
    } else if (segunda < 0) {
        console.log(`x = ${x}: Máximo local`);
    } else {
        console.log(`x = ${x}: Ponto de inflexão ou indeciso`);
    }
});